edx/edx-enterprise
# [ENT-105] Enroll PendingEnterpriseCustomerUsers in courses when converted

This pull request adds a feature such that when learners are enrolled in a course via the Manage Learners pane, but they don't exist, and as such are created as `PendingEnterpriseCustomerUser`s, a `PendingEnrollment` object is created that enrolls the user in a course when the learner creates an actual platform account.

**JIRA tickets**: Implements ENT-105

**Dependencies**: None

**Sandbox URL**: https://ent105.opencraft.hosting/ (if DNS hasn't shown up, use direct IP address: http://213.32.76.186; use `staff` account for superuser access)

**Merge deadline**: 2017-01-03

**Testing instructions**:

1. Create an `EnterpriseCustomer`
2. Create a course, noting its course ID (including the `course-v1` prefix)
3. In the EnterpriseCustomer's Manage Learners page, attempt to add a learner with an email address that does not yet exist as a user in the platform. Be sure to also attempt to enroll it in the course.
4. Note that a PendingEnterpriseCustomerUser object is created.
5. Using the standard registration methods, create a user with the aforesaid email address.
6. Note that the user is immediately enrolled in the course with the appropriate mode without further intervention.

**Author notes and concerns**:

1. At present, it was necessary to do the registration with local database methods; the entire user post-save method is wrapped in a single transaction; if we do an HTTP API call, then the user doesn't exist yet (from the perspective of the API) in order for the enrollment to be processed. In Django 1.9 and later, a `post_commit` hook is available, which will allow us to bypass that by making the enrollment HTTP call after the database commit is finalized.

**Reviewers**

**Merge checklist:**

**Post merge:**
      finished.
